### DISCUSSION TITLE
KeyClu & skhd

### DISCUSSION THREAD

**Discussion by gldtn** [Anze/KeyCluCask#46] [Ideas]
Can I list my own shortcuts I have inside my skhdrc?

Basically I would like to press a keybind and have it show all my shortcuts I have listed in my skhd. Is this possible, even if I have to manually add them.

I guess what I'm asking for is the ability to add custom sheets and assign to a key binding in skhd
hyper - s : KeyClu --sheet SHEETNAME --toggle-persistent 

This would be a great feature for programs like Yabai/skhd and other global shortcuts not related to s specific program!

Of course if something similar doesn't already exist.

Thanks!

**Comment by Anze** [comment]
Hey [gldtn](https://github.com/gldtn),
That's an interesting feature that could be helpful. However, there's a small issue: in the configuration, only the shortcut and its assigned action are present, lacking a description to explain what it is. I believe this can be resolved by adding additional comments to each shortcut, providing a brief description of the corresponding action.

Like:
```
#KeyClu: Do some magic
hyper - r: Atom.app . 
```
Than shortcut can be resolved to:
```
Do some magic       ⌃⌥⌘⇧ R
```

What do you think?

**Comment by gldtn** [comment]
That would be perfect, I already add a comment on most of my keybinds inside skhd, so all I'll need to do is add `#KeyClu:` to them. If it's not too difficult, the ability to be able to add them to groups as well..

```
#KeyClu[Space management]: Change/focus space to space #1
alt - 1 : yabai -m space --focus 1
#KeyClu[Space management]: Change/focus space to space #2
alt - 2 : yabai -m space --focus 2

#KeyClu[Window management]:  Move/Focus window west
alt - h : yabai -m window --focus west  || yabai -m display --focus west
#KeyClu[Window management]:  Move/Focus window east
alt - l : yabai -m window --focus east  || yabai -m display --focus east
```

Just a thought to be able to organize it in the sheet, I'm assuming that's what you meant I'll need to add #keyClu: Description before each shortcut. If not possible, the above implementation as you suggested would be awesome.

If you add this as a pro feature, I would be happy to pay for it. KeyClu deserves some love ;)

Thank you!

**Comment by gldtn** [comment]
Hey @Anze, just checking in to see if you had any more ideas towards this request.

Thanks!

**Comment by Anze** [comment]
Hey [gldtn](https://github.com/gldtn),
Recently introduced listing skhd shortcuts in v0.22. Details of the current implementation can be found in the [wiki](https://github.com/Anze/KeyCluCask/wiki/Integrations-%E2%80%90-skhd) page.
Would love to hear your thoughts on it.

**Comment by gldtn** [comment]
Thanks for implementing this request!

I have my skhdrc setup, and enabled in KeyClu settings, how do I get it to show the shortcuts sheet for skhd? To I need pass any parameters for it to show up?

**Comment by Anze** [comment]
Hey [gldtn](https://github.com/gldtn),
When you activate Shortcuts Panel, the skhd shortcuts currently appears in the last column for any app.

**Comment by gldtn** [comment]
I might be doing something wrong because I don't see any skhd shortcuts at all. So under integration I enabled skhd, and I simplified my skhdrc to make sure there was nothing preventing it from reading it.

So I tried it with this and even only one item. I restarted KC after the changes and no luck on my end..
```shell
# Change/focus space 1
alt - 1 : yabai -m space --focus 1

#KeyClu[Window Management]: Change/focus space 2
alt - 2 : yabai -m space --focus 2
```

It doesn't show neither under --show-shortcuts or the persistent panel.

**Reply by Anze** [reply]
Can you please check that `.skhdrc` file placed at `/Users/<username>/.skhdrc`?
Update: looks like i've missed that there might be other locations for skhd config file. Would address this issue in the upcoming release.

**Reply by gldtn** [reply]
Yes, that was the problem as mine is located in ~/.config/skhd/skhdrc

I'll play with it throughout the weekend and let you know how it goes. A quick thought, is it possible to have separate sheet just for skhdrc and call it with a shortcut within skhdrc? 

Thanks for the implementation!

**Reply by Anze** [reply]
Yeah, it's doable.
When I was thinking it through, I had some questions like: how should it look, whether there's a little number or a lots of items? The hiccup is that right now the window size doesn't flex with the number of items. Would try to come up with a solution. Or might you have something in mind already, like how it should looks like?

**Reply by gldtn** [reply]
I see what you mean.. most SKHD user have quite a lot of shortcuts items. I'm still having a bit of difficulty organizing mine because all the items I have, here is my .skhdrc file so you have an idea: https://github.com/gldtn/dotfiles/blob/main/.config/skhd/skhdrc

**Reply by Anze** [reply]
Thanks for sharing the file sample. It's been quite helpful. I've found and resolved a few issues with the format parser.

**Reply by Anze** [reply]
Hey [gldtn](https://github.com/gldtn),
I've added panel to present skhd shortcuts, to get it pass `--toggle-panel --sheet skhd` params to KeyClu.

**Reply by gldtn** [reply]
Looks awesome, once again thanks for implementing this request. Can't thank you enough sending a nice cup of coffee on the way..

I had some time to play with it today. Couple of things I came up with and noticed a small bug;

- Shortcuts shows in a random order every time I save skhd. Try simply saving and calling  the sheet, you'll see what I mean. It changes every time the file is saved.
- If doable.. would be nice to maybe be able to order how they appear, or follow skhdrc order from top to bottom.
- Maybe an option to set number of columns `--toggle-panel --col 3 --sheet skhd`?

- Regular/Common shortcuts seems to have a bug, its hiding the first item

See my updated skhd file: https://github.com/gldtn/dotfiles/blob/main/.config/skhd/skhdrc

Here is the exact commented line: https://github.com/gldtn/dotfiles/blob/8a7d896fa6c01ae0fa6c10413aa82dcc1c65b92c/.config/skhd/skhdrc#L169

**Reply by Anze** [reply]
Appreciate you taking the time to check and share your feedback. I'll keep in mind your suggestions.
As for now I've made a quick fix build, where addressed issues with hiding first item in commons and random order of groups, now its alphabetical order. Also a bit adjusted how amount of columns is defined.
While it is a quick fix, the complete solution would be in the upcoming release.

**Reply by gldtn** [reply]
Hiding first item and random sorting is fixed. Thanks for the quick fix!

**Comment by Anze** [comment]
Hey [gldtn](https://github.com/gldtn),
I just made new release, there few improvements made on skhd. Also added optional param to specify columns amount. Details described in the [wiki page](https://github.com/Anze/KeyCluCask/wiki/Integrations-%E2%80%90-CLI-params).

**Reply by gldtn** [reply]
The columns amount param is working nicely, this is a great improvement for those with a lot of shortcuts. Although the alphabetical order of items seems to have stopped working.

Thanks once again for the updates!

**Reply by Anze** [reply]
Hey [gldtn](https://github.com/gldtn),
Thanks. I've changed alphabetical order to order how it is defined in file, with the exception of "Common," which should remains the first entry, with some limits. This adjustment provide better flexibility to maintain order by yourself.
